### PR TITLE
Bump to xamarin/java.interop/master@dcaf794d

### PR DESCRIPTION
**Pull request by jonpryor**
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1139578

Context: https://github.com/xamarin/java.interop/pull/691
Context: https://liquid.microsoft.com/Web/Object/Read/ms.security/Requirements/Microsoft.Security.SystemsADM.10039#guide

Changes: https://github.com/xamarin/java.interop/compare/007b35b489173010de038c95cb2a0d3ec514f30d...dcaf794d26206834a0b08bda85ba85b1c15b7bd1

  * xamarin/Java.Interop@dcaf794d: [generator] Fix error and warning codes values (#697)
  * xamarin/Java.Interop@08ff4db1: [java-interop, Java.Interop] Securely load native libs (#691)
  * xamarin/Java.Interop@09c68911: [generator] Clean up generated whitespace (#692)
  * xamarin/Java.Interop@d664e90e: [generator] Make warning and error messages localizable. (#689)

For improved security, the guidance is that when loading `.dll` files
on Windows, the default [`LoadLibrary()`][0] behavior is to be
*avoided*, as the [Dynamic-Link Library Search Order][1] includes the
"current working directory" and directories listed in `%PATH%`, both
of which are under control of an "attacker" and not the application.

Instead, [`LoadLibraryEx()`][2] should be used, providing a set of
`LOAD_LIBRARY_SEARCH_*` values which constrain the set of directories
that will be searched for the library (if a full path is not used)
and/or the library's dependencies.

Historically, Xamarin.Android hasn't directly called any
`LoadLibrary()` function.  Instead, xamarin-android called
[**dlopen**(3)][3], then used dlfcn-win32/dlfcn-win32@ef7e412d to
implement `dlopen()` in terms of `LoadLibraryEx()`.

Unfortunately, dlfcn-win32/dlfcn-win32@ef7e412d calls `LoadLibraryEx()`
with `LOAD_WITH_ALTERED_SEARCH_PATH`, which while a security
improvement, does not go far enough for our internal requirements.

Migrate away from dlfcn-win32 and instead use the new
`java-interop-dlfcn.h` family of functions added in
xamarin/java.interop@08ff4db1.  These new functions appropriately use
`LoadLibraryEx()` with a constrained set of `LOAD_LIBRARY_SEARCH_*`
flags.

*Non-obvious semantic note*: On Mono for all platforms, P/Invoking
`__Internal` is equivalent to trying to load the path `NULL`, e.g.
on macOS `dlopen(NULL, …)`.  However, only Android allows a custom
"`loader_func`" to intercept the attempted load of `NULL` and "do
something" with it, which xamarin-android *does* intercept, "remapping"
`NULL` to `libxa-internal-api.so`.  On all other platforms,
`loader_func` cannot intercept this invocation.

…which made for an "interesting" interaction: on macOS, if we used:

	api_dso_handle = java_interop_lib_load (dso_path.get (), JAVA_INTEROP_LIB_LOAD_LOCALLY, nullptr);

then the macOS Designer integration tests would fail!

	Renderer >> 4 [monodroid] Calling into managed runtime init
	Renderer (error) >>
	Renderer (error) >> Unhandled Exception:
	Renderer (error) >> System.EntryPointNotFoundException: java_interop_jnienv_get_java_vm assembly:<unknown assembly> type:<unknown type> member:(null)
	Renderer (error) >> at (wrapper managed-to-native) Java.Interop.NativeMethods.java_interop_jnienv_get_java_vm(intptr,intptr&)
	Renderer (error) >> at Java.Interop.JniEnvironment+References.GetJavaVM (System.IntPtr jnienv, System.IntPtr& vm) [0x00000] in <0f003a4904fd44d0a8cc6a63962ab40b>:0
	Renderer (error) >> at Java.Interop.JniEnvironmentInfo.set_EnvironmentPointer (System.IntPtr value) [0x00037] in <0f003a4904fd44d0a8cc6a63962ab40b>:0
	Renderer (error) >> at Java.Interop.JniEnvironmentInfo..ctor (System.IntPtr environmentPointer, Java.Interop.JniRuntime runtime) [0x00006] in <0f003a4904fd44d0a8cc6a63962ab40b>:0
	Renderer (error) >> at Java.Interop.JniRuntime..ctor (Java.Interop.JniRuntime+CreationOptions options) [0x0017b] in <0f003a4904fd44d0a8cc6a63962ab40b>:0
	Renderer (error) >> at Android.Runtime.AndroidRuntime..ctor (System.IntPtr jnienv, System.IntPtr vm, System.Boolean allocNewObjectSupported, System.IntPtr classLoader, System.IntPtr classLoader_loadClass, System.Boolean jniAddNativeMethodRegistrationAttributePresent) [0x00000] in /Users/builder/azdo/_work/4/s/xamarin-android/src/Mono.Android/Android.Runtime/AndroidRuntime.cs:25

In order for the macOS Designer integration tests to succeed,
`libxa-internal-api.dylib` *must* be loaded as `RTLD_GLOBAL`, which is
done via `JAVA_INTEROP_LIB_LOAD_GLOBALLY`.

[0]: https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya
[1]: https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order
[2]: https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa
[3]: https://linux.die.net/man/3/dlopen

### CODE REVIEW

**Comment by jonpryor** [comment]
Related: https://github.com/xamarin/java.interop/pull/691

File: src/monodroid/jni/monodroid-glue.cc [hunk 1]
```diff
@@ -10,7 +10,10 @@
 #include <errno.h>
 #include <limits.h>
 
+#if defined (APPLE_OS_X)
```
Comment (grendello) [inline src/monodroid/jni/monodroid-glue.cc:13]: What about Linux? :)

**Review by grendello — CHANGES_REQUESTED** [review]

File: src/monodroid/jni/monodroid-glue.cc (hunk 1 above)
Comment (jonpryor) [reply src/monodroid/jni/monodroid-glue.cc:13]: The only reason it's needed *at all* is for `MonodroidRuntime::get_my_location()`, which on macOS uses `dladdr()`.  Linux doesn't use `dladdr()`, and thus it isn't needed.

**Review by radekdoulik — COMMENTED** [review]
There is something wrong with warning codes, like in `Xamarin.Android.Build.Tests.BuildTest.GeneratorValidateEventName(False,True,"","")` test. Where we expect `BG8405` warning, but we get `BG2138`:

```
  BINDINGSGENERATOR : warning BG2138: Event name for 'Com.Xamarin.Testing.Test.IOn123Listener.SetOn123Listener' is invalid. A valid 'eventName' or 'argsType' can be assigned by adding a rule to the Metadata.xml transforms file. [/Users/runner/work/1/s/bin/TestRelease/temp/GeneratorValidateEventNameFalseTrue/UnnamedProject.csproj]
```

The `Localized` message seems to have the right number:
```
public static LocalizedMessage WarningInvalidEventName => new LocalizedMessage (8504, Java.Interop.Localization.Resources.Generator_BG8504);
```

@jpobst could you please take a look?

**Comment by radekdoulik** [comment]
Oh, I see where is the problem. We print it as hexadecimal value, which I guess is not what we want. I will try to fix it.

**Comment by radekdoulik** [comment]
We have some hexadecimal coded warnings, so I ended up with making the codes as hexdecimal values as well, so that we can print everything as before the localization changes.

This PR https://github.com/xamarin/java.interop/pull/697 should fix it and we can update this PR once it is merged.

**Review by radekdoulik — APPROVED** [review]

**Comment by jonpryor** [comment]
Forgot to also mention: https://github.com/mono/mono/issues/20295

**Pull request merged**